galaxyproject/galaxy
# Add OIDC PKCE parameter support

Relevant Issue: https://github.com/galaxyproject/galaxy/issues/13605

This pr is to add Proof Key for Code Exchange (PKCE) support when using OIDC auth. Essentially, it's a few extra parameters added to some OAuth requests for extra security. Some OIDC providers recommend having support for these parameters, and it may be required for some. 

I tested this pr manually with Keycloak using a setup that required PKCE parameters, and then both a Keycloak setup and a Google setup  that did not require them making sure existing functionality was not affected. I also added a unit test with the other OIDC auth tests that followed the pattern of other already existing Galaxy unit tests in `test_custos_authnz.py`. 

I did add a file called `pkce_utils.py` in `galaxy/lib/galaxy/authnz`. This file is just taken from a python package (https://pypi.org/project/pkce/), and I included the license from that, because I wasn't sure how to handle the dependency management.. Let me know if this would need to be changed. I did not make changes to any of the docs. This is my first pr to this project, so let me know if I need to change anything.

The spec for PKCE is here: https://datatracker.ietf.org/doc/html/rfc7636
More info can be found here: https://www.oauth.com/oauth2-servers/pkce/


## How to test the changes?
(Select all options that apply)
- [x ] Instructions for manual testing are as follows:
  1. Make sure the OIDC support is enabled in `galaxy.yml`. Instructions are found here: https://galaxyproject.org/authnz/config/oidc/#enable-oidc-based-login 
  2. For testing with Keycloak, in `oidc_backends_config`, under the Keycloak configuration add a xml attribute called 
`pkce_support` like the following to an existing Keycloak config:
  
 ```oidc_backends_config.xml

    <provider name="Keycloak">
        
        <url>http://localhost-uri:8080/auth/realms/master/</url>
        <pkce_support>true</pkce_support>
         ...
 ```
  3. Configure Keycloak to use PKCE. Instructions can be found in the Keycloak docs by searching for PKCE: https://www.keycloak.org/docs/latest/server_admin/#_oidc_clients. 
  4. Start Galaxy, and open up the home page.
  5. Click on the login link in the top right.
  6. Choose to sign in with KeyCloak. 
  7. Sign in as a Keycloak user.
  8. You should be redirected back to the Galaxy homepage and signed in.
  
 

## License
